File: leetcode/0198-house-robber/solution.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums[0], nums[1])
        
        cache = list()
        cache.append(nums[0])
        cache.append(max(nums[0], nums[1]))
        
        for i in range(2, len(nums)):
            cache.append(max(nums[i] + cache[i - 2], cache[i - 1]))
        
        return cache[len(nums) - 1]

File: leetcode/0198-house-robber/solution.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
         if (nums.size() == 1)
            return nums[0];
        if (nums.size() == 2)
            return max(nums[0], nums[1]);
        
        int cache[nums.size()];
        cache[0] = nums[0];
        cache[1] = max(nums[0], nums[1]);
        
        for (int i = 2; i < nums.size(); i++)
            cache[i] = max(cache[i - 1], nums[i] + cache[i - 2]);
        
        return cache[nums.size() - 1];
    }
};